cunumeric/module.py

```diff
@@ -714,10 +696,25 @@ def _contract(expr, a, b=None, out=None, stacklevel=1):
     if len(set(out_modes)) != len(out_modes):
         raise ValueError("Duplicate mode labels on output tensor")
 
-    # TODO: Handle duplicate modes on inputs
-    if len(set(a_modes)) != len(a_modes) or len(set(b_modes)) != len(b_modes):
-        raise NotImplementedError("Duplicate mode labels on input tensor")
-
+    # Handle duplicate modes on inputs
+    c_a_modes = Counter(a_modes)
+    for mode in c_a_modes:
+        count = c_a_modes[mode]
+        if count > 1:
+            axes = [i for (i, m) in enumerate(a_modes) if m == mode]
+            a = a.diag_helper(axes=axes)
+            # diagonal is stored on last axis
+            for _ in range(count - 1):
+                a_modes.remove(mode)
```

Comment:
Since diag_helper now always puts the diagonal dimension last, we need to change this to:
```
a_modes = [m for m in a_modes if m != mode] + [mode]
```
and similar for b